engine/src/main/coffee/serialize/readexportedvalue.coffee

```diff
@@ -0,0 +1,247 @@
+# (C) Uri Wilensky. https://github.com/NetLogo/Tortoise
+
+{ BreedNamePair
+, BreedReference
+, ExportedCommandLambda
+, ExportedLinkSet
+, ExportedPatchSet
+, ExportedReporterLambda
+, ExportedTurtleSet
+, LinkReference
+, NobodyReference
+, PatchReference
+, TurtleReference
+} = require('./exportstructures')
+
+{ fold, isSomething, map: mapMaybe, maybe, None } = require('brazier/maybe')
+
+# (String) => Number
+firstIndexOfUnescapedQuote = (str) ->
+  index = str.indexOf('"')
+  if index > 0
+    if str[index - 1] isnt "\\"
+      index
+    else
+      1 + index + firstIndexOfUnescapedQuote(str.slice(index + 1))
+  else
+    index
+
+# (String, (String) => Any, (String) => Any) => Array[Any]
+parseList = ->
+
+  parseListHelper = (list, readValue, readAgentLike) ->
+
+    parseInner = (contents, acc = [], accIndex = 0) ->
+
+      strIndex = ( char) -> contents.indexOf(char)
+      strFrom  = (index) -> contents.slice(index)
+      strUntil = (index) -> contents.slice(0, index)
+      tempered = (index) -> index + (if contents[index + 1] is ']' then 1 else 2)
+
+      recurse = (nextIndex, item) ->
+        parseInner(strFrom(nextIndex), acc.concat([item]), accIndex + nextIndex)
+
+      if not (contents.startsWith('(anonymous command:') or contents.startsWith('(anonymous reporter:'))
+        switch contents[0]
+          when ']' # End of list
+            [acc, accIndex + 1]
+          when '[' # Start of list
+            [item, endIndex] = parseListHelper(contents, readValue, readAgentLike)
+            recurse(tempered(endIndex), item)
+          when '{' # Start of agent/agentset
+            index = strIndex('}')
+            recurse(tempered(   index), readAgentLike(strUntil(index + 1)))
+          when '"' # Start of string
+            index = firstIndexOfUnescapedQuote(strFrom(1)) + 1
+            recurse(tempered(   index), readValue(    strUntil(index + 1)))
+          else
+            rightIndex = strIndex(']') # End of next item, if there's no item after it
+            spaceIndex = strIndex(' ') # Separator between next item and the one after
+            if rightIndex < spaceIndex or spaceIndex < 0
+              recurse(rightIndex    , readValue(strUntil(rightIndex)))
+            else
+              recurse(spaceIndex + 1, readValue(strUntil(spaceIndex)))
+      else
+        throw new Error("Importing a list of anonymous procedures?  Not happening!")
```

Comment:
Does NetLogo Desktop actually export lists of anonymous procedures? If it does, that seems like a bug to me.

Reply:
It actually does not.  Instead, it will complain, "Expected a literal value", and import the value as `0`.  It does this both for anonprocs in lists and the top-level anonprocs.

My approach has been this instead:

  * At the top-level, I'll import the anonymous procedure... kind of.  I'll at least put an anonymous procedure where you imported one, but if you try to run it, the *execution* of the anonproc raises an error.
  * I don't really want to write a sophisticated list-parser (as evidenced above), so, if I'm parsing a list and I find an anonproc, I throw an error immediately.

Fortunately, this almost never comes up (with BeeSmart Hive Finding and State Machine Example being the only two models (of ~90) in our test suite that even store anonprocs in variables).

I'm open to doing things a bit differently in Tortoise than they're currently implemented here.  I'm not sure that I could perfectly replicate the error messages that desktop gives off, though.  Its error messages report the line of the CSV file where the error occurred, and I don't really have that information available to me at this stage in the parsing, and I'm not sure that I could reasonably get it.

Reply:
Yeah, that seems like a good plan. It sounds like desktop should maybe change to avoid exporting anonymous procedures.

Reply:
I would recommend against that for two reasons:

  1. Supposedly, some people want to `export-world` so they can analyze data about their model outside of NetLogo.  Such a change would block them from being able to do so if they had anonprocs in variables, when those anonprocs are causing no actual problems for these people.
  2. I don't have evidence that such a model exists, but it's not hard for me to imagine that some model would have anonprocs in variables that it isn't necessarily going to use in the course of running the model after imported.  I think we have numerous models where globals are only used when the model in running with a particular configuration (e.g. particular Switch enabled, or Chooser option selected).  (PNoM models come to mind.)  So, since the "import" isn't necessarily going to be harmful, I'm inclined to gives the lambdas a pass until the point at which they become problematic.

Reply:
That all makes sense too. I guess no changes are needed then. Thanks for investigating all of this so thoroughly!